lib/slackUtils.ts

```diff
@@ -1,224 +1,90 @@
-			return members;
-		} catch (error) {
-			console.error(`Error loading members for team ${token.team_name}:`, error);
-			return [];
-		}
-	})).then((usersArray) => {
-		loadMembersDeferred.resolve(flatten(usersArray));
-	});
-
-	Promise.all(tokens.map(async (token) => {
-		try {
-			const {emoji}: any = await webClient.emoji.list({token: token.access_token});
-			const {team}: any = await webClient.team.info({token: token.bot_access_token});
-			return Object.entries(emoji).map(([name, url]) => ({
-				team: team.id,
-				name,
-				url,
+import SlackCache from './slackCache';
+
+const slackCaches = new Map<string, SlackCache>();
+const initializedSlackCachesDeferred = new Deferred<void>();
+
+// Immediately Invoked Function Expression
+(function initializedSlackCaches() {
+	getTokens().then(async (tokens) => {
+		for (const token of tokens) {
+			const rtmClient: RTMClient = await getRtmClient(token.team_id);
+			slackCaches.set(token.team_id, new SlackCache({
+				token,
+				rtmClient,
+				webClient: new WebClient,
+				enableUsers: true,
+				enableEmojis: true,
+				enableReactions: token.team_id === process.env.TEAM_ID,
 			}));
-		} catch (error) {
-			console.error(`Error loading emojis for team ${token.team_name}:`, error);
-			return [];
 		}
-	})).then((emojisArray) => {
-		loadEmojisDeferred.resolve(flatten(emojisArray));
+		initializedSlackCachesDeferred.resolve();
 	});
-});
+})();
 
+// Note: support only TSG
 export const getReactions = async (channel: string, ts: string) => {
-	// reaction_addedイベントの中から呼ばれても安全なように常にnextTickを待つ
-	await new Promise(process.nextTick);
-
-	const key = `${channel}\0${ts}`;
-	if (reactionsCache.has(key)) {
-		return reactionsCache.get(key);
-	}
-
-	const tokens = await tokensDeferred.promise;
-	const token = tokens.find((token) => token.team_id === process.env.TEAM_ID);
-	if (!token) {
-		throw new Error('Available token not found');
-	}
-
-	const data = await webClient.conversations.history({
-		token: token.access_token,
-		channel: channel,
-		latest: ts,
-		limit: 1,
-		inclusive: true,
-	});
-
-	// race condition
-	if (reactionsCache.has(key)) {
-		return reactionsCache.get(key);
-	}
-
-	const remoteReactions = get(data, ['messages', 0, 'reactions'], [] as Reaction[]);
-	const remoteReactionsObj = Object.fromEntries(remoteReactions.map((reaction) => (
-		[reaction.name, reaction.count]
-	)));
-	reactionsCache.set(key, remoteReactionsObj);
-	return remoteReactionsObj;
+	await initializedSlackCachesDeferred.promise;
+	return slackCaches.get(process.env.TEAM_ID).getReactions(channel, ts);
 };
 
 export const getAllTSGMembers = async (): Promise<Array<any>> => {
-	return [
-		...additionalMembers,
-		...(await loadMembersDeferred.promise),
-	].filter(user => user.team_id === process.env.TEAM_ID);
+	await initializedSlackCachesDeferred.promise;
+	return Array.from(await slackCaches.get(process.env.TEAM_ID).getUsers());
```

Comment:
これ process.env.TEAM_ID は string | undefined になってると思うんですけど、! なしでコンパイル通りますかね? (チェックしてなくてすまん)

Reply:
https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#null-and-undefined
> By default, the type checker considers null and undefined assignable to anything. Effectively, null and undefined are valid values of every type. That means it’s not possible to stop them from being assigned to any type, even when you would like to prevent it. The inventor of null, Tony Hoare, calls this his [“billion dollar mistake”](https://wikipedia.org/wiki/Null_pointer#History).
> The [strictNullChecks](https://www.typescriptlang.org/tsconfig#strictNullChecks) flag fixes this: when you declare a variable, it doesn’t automatically include null or undefined. You can include them explicitly using a union type:

strictNullChecksってのをtsconfigでONにしないと通るらしいです

Reply:
うーん、ほんとにこれかな

Reply:
うん、これだ

Reply:
え、TypeScriptめんどくさ　やめましょう

Reply:
ぜぇぜぇ c7b0f5c